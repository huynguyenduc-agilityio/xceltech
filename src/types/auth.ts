export interface ILoginUser {
  email: string;
  password: string;
}

export interface IAuthUser {
  refresh: string;
  access: string;
  user: {
    id: string;
    email: string;
  };
}

export interface IRegisterUser extends ILoginUser {
  firstName: string;
  lastName: string;
  phone: string;
  confirmPassword: string;
  isAgreeTerms?: boolean;
  isReceiveNewsletters?: boolean;
}

export interface IActiveUser {
  uidb64: string | null;
  token: string | null;
}
