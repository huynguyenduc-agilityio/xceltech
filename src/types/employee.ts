import { z } from 'zod';

import {
  educationSchema,
  familySchema,
  financialSchema,
  guarantorSchema,
} from '@/utils';

export type EducationFormValues = z.infer<typeof educationSchema>;
export type GuarantorFormValues = z.infer<typeof guarantorSchema>;
export type FamilyFormValues = z.infer<typeof familySchema>;
export type FinancialFormValues = z.infer<typeof financialSchema>;

export interface IEmployeeEducationInfo extends EducationFormValues {
  id?: string;
  startDate?: Date | string;
  endDate?: Date | string;
  type?: EducationType;
}

export enum EducationType {
  Academic = 'Academic',
  Professional = 'Professional',
}

export interface IEmployeeGuarantor extends GuarantorFormValues {
  id?: string;
}

export interface IEmployeeFamilyInfo extends FamilyFormValues {
  id?: string;
}

export interface IEmployeeFinancialInfo extends FinancialFormValues {
  id?: string;
}

export type DocumentsForm = {
  documents: {
    offerLetter?: File | null;
    birthCertificate?: File | null;
    guarantorForm?: File | null;
    degreeCertificate?: File | null;
  };
};

export type Documents = {
  documentFile: string;
  documentType: string;
};
