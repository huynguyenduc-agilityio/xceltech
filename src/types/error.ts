export type ErrorType = {
  detail: string;
};
