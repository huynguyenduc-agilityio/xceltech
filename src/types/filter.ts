export interface FilterCriteria {
  [key: string]: string | number | boolean | string[];
}
