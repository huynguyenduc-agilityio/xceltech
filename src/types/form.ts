export interface UploadFileForm {
  [key: string]: File | null;
}

export type TField<T> = {
  name: keyof T;
  label: string;
  placeholder: string;
  min?: string;
  max?: string;
  rules: {
    required: string;
    pattern: {
      value: RegExp;
      message: string;
    };
  };
  ariaLabel: string;
};

export enum MutationType {
  Create = 'create',
  Edit = 'edit',
}
