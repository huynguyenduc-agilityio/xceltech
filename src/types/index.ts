export * from './status';
export * from './employee';
export * from './table';
export * from './auth';
export * from './form';
export * from './tab';
export * from './user';
export * from './queryKey';
export * from './leave';
export * from './filter';
export * from './error';
export * from './notification';
export * from './profile';
