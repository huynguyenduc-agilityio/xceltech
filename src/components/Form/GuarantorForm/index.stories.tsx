import type { Meta, StoryObj } from '@storybook/react';

import { MutationType } from '@/types';

// Components
import GuarantorForm from '.';

const meta = {
  title: 'Components/GuarantorForm',
  component: GuarantorForm,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  render: (args) => (
    <div className="flex items-center justify-center w-screen py-20 bg-blue-light">
      <div className="bg-white w-[988px] rounded-[15px]">
        <div className="p-10">
          <GuarantorForm {...args} />
        </div>
      </div>
    </div>
  ),
} satisfies Meta<typeof GuarantorForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    mode: MutationType.Create,
  },
};
