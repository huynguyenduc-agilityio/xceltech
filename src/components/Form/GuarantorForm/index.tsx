import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';

// Types
import {
  IEmployeeGuarantor,
  MutationType,
  ToastStatus,
  GuarantorFormValues,
} from '@/types';

// Components
import { Button, Form, FormField, TextField } from '@/components';

// Constants
import { BASE_GUARANTOR_FIELD, MESSAGES } from '@/constants';

// Utils
import { guarantorSchema } from '@/utils/schemas/employee';

// Hooks
import { useGuarantorMutation, useToast } from '@/hooks';

export interface IGuarantorForm {
  mode: MutationType;
  initialValues?: Partial<IEmployeeGuarantor>;
  onBack?: () => void;
}

const GuarantorForm = ({
  initialValues = {},
  mode,
  onBack,
}: IGuarantorForm) => {
  const { toast } = useToast();
  const { handleGuarantorMutation, isGuarantorMutationLoading } =
    useGuarantorMutation({
      type: mode,
    });

  const { name = '', job = '', phone = '' } = initialValues || {};
  const defaultValues: IEmployeeGuarantor = {
    name,
    job,
    phone,
  };

  const form = useForm<GuarantorFormValues>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    resolver: zodResolver(guarantorSchema),
    defaultValues,
  });

  const {
    control,
    handleSubmit,
    reset,
    getValues,
    formState: { isValid, isDirty, isSubmitting },
  } = form;

  const isCreate = mode === MutationType.Create;

  const onSubmit = async (data: IEmployeeGuarantor) => {
    const payload = {
      ...data,
      ...(!isCreate && initialValues?.id ? { id: initialValues.id } : {}),
    };

    try {
      await handleGuarantorMutation(payload);

      toast({
        status: ToastStatus.Success,
        title: isCreate
          ? MESSAGES.COMMON.ADD_SUCCESS('Guarantor')
          : MESSAGES.COMMON.UPDATE_SUCCESS('Guarantor'),
      });
      onBack?.();

      reset(initialValues);
    } catch {
      toast({
        status: ToastStatus.Error,
        title: isCreate
          ? MESSAGES.COMMON.ADD_FAILED('Guarantor')
          : MESSAGES.COMMON.UPDATE_FAILED('Guarantor'),
      });

      reset(getValues());
    }
  };

  const disableSubmit = !isDirty || !isValid || isSubmitting;

  return (
    <Form {...form}>
      <form onSubmit={handleSubmit(onSubmit)} className="">
        <div className="w-full text-black-default space-y-10">
          {BASE_GUARANTOR_FIELD.map(({ name, label, placeholder }) => (
            <FormField
              key={name}
              control={control}
              name={name}
              render={({ field, fieldState: { error } }) => (
                <TextField
                  label={label}
                  variant="primary"
                  className="space-y-5"
                  labelClassName="text-md"
                  inputClassName="h-[68px] px-12 rounded-[15px]"
                  placeholder={placeholder}
                  {...field}
                  errorMessage={error?.message}
                />
              )}
            />
          ))}

          <div className="w-full h-[70px]">
            <Button
              type="submit"
              className="w-[364px] bg-green-primary"
              isLoading={isGuarantorMutationLoading}
              disabled={disableSubmit}
            >
              {isCreate ? 'Submit' : 'Update'}
            </Button>
          </div>
        </div>
      </form>
    </Form>
  );
};

export default GuarantorForm;
