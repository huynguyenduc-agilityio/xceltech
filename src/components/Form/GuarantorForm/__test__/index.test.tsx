import { fireEvent, render, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

import { MutationType } from '@/types';

// Components
import GuarantorForm from '..';

const renderComponent = () => {
  const queryClient = new QueryClient();

  return render(
    <QueryClientProvider client={queryClient}>
      <GuarantorForm mode={MutationType.Edit} />
    </QueryClientProvider>,
  );
};

describe('GuarantorForm Component', () => {
  it('should match snapshot', () => {
    const container = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it('handle input value change and form submit', () => {
    const { getByPlaceholderText, getByRole } = renderComponent();
    const jobTitleInput = getByPlaceholderText(
      'Enter your job title / occupation',
    );
    const phoneInput = getByPlaceholderText('Enter your phone number');
    const nameInput = getByPlaceholderText('Enter your guarantor’s name');
    const submitButton = getByRole('button', { name: 'Update' });

    fireEvent.change(jobTitleInput, { target: { value: 'Developer' } });
    fireEvent.change(phoneInput, { target: { value: '08123456789' } });
    fireEvent.change(nameInput, { target: { value: 'John Doe' } });

    fireEvent.click(submitButton);

    expect(jobTitleInput).toHaveValue('Developer');
    expect(phoneInput).toHaveValue('08123456789');
    expect(nameInput).toHaveValue('John Doe');
  });

  it('should show error message fields', async () => {
    const { getByPlaceholderText, getByText } = renderComponent();

    const jobTitleInput = getByPlaceholderText(
      'Enter your job title / occupation',
    );
    const phoneInput = getByPlaceholderText('Enter your phone number');
    const nameInput = getByPlaceholderText('Enter your guarantor’s name');

    fireEvent.change(jobTitleInput, { target: { value: '' } });
    fireEvent.change(phoneInput, { target: { value: '' } });
    fireEvent.change(nameInput, { target: { value: '' } });

    fireEvent.blur(jobTitleInput);
    fireEvent.blur(phoneInput);
    fireEvent.blur(nameInput);

    await waitFor(() => {
      expect(
        getByText('Job Title / Occupation is required!'),
      ).toBeInTheDocument();
      expect(getByText('Phone No is required!')).toBeInTheDocument();
      expect(getByText('Guarantor name is required!')).toBeInTheDocument();
    });
  });
});
