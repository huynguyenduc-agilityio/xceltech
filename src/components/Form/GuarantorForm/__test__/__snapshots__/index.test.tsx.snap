// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuarantorForm Component should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class=""
      >
        <div
          class="w-full text-black-default space-y-10"
        >
          <div
            class="grid w-full gap-2"
          >
            <div
              class="w-full space-y-5"
            >
              <label
                class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                id="Guarantor’s Name-label"
              >
                Guarantor’s Name
              </label>
              <input
                aria-describedby=":r0:-form-item-description"
                aria-invalid="false"
                aria-labelledby="Guarantor’s Name-label"
                autocomplete="off"
                class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                data-testid="input-element"
                id=":r0:-form-item"
                name="name"
                placeholder="Enter your guarantor’s name"
                value=""
              />
            </div>
          </div>
          <div
            class="grid w-full gap-2"
          >
            <div
              class="w-full space-y-5"
            >
              <label
                class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                id="Job Title / Occupation-label"
              >
                Job Title / Occupation
              </label>
              <input
                aria-describedby=":r1:-form-item-description"
                aria-invalid="false"
                aria-labelledby="Job Title / Occupation-label"
                autocomplete="off"
                class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                data-testid="input-element"
                id=":r1:-form-item"
                name="job"
                placeholder="Enter your job title / occupation"
                value=""
              />
            </div>
          </div>
          <div
            class="grid w-full gap-2"
          >
            <div
              class="w-full space-y-5"
            >
              <label
                class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
                id="Phone No-label"
              >
                Phone No
              </label>
              <input
                aria-describedby=":r2:-form-item-description"
                aria-invalid="false"
                aria-labelledby="Phone No-label"
                autocomplete="off"
                class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
                data-testid="input-element"
                id=":r2:-form-item"
                name="phone"
                placeholder="Enter your phone number"
                value=""
              />
            </div>
          </div>
          <div
            class="w-full h-[70px]"
          >
            <button
              class="font-productsans inline-flex items-center justify-center whitespace-nowrap rounded-[14px] text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed text-white disabled:hover:bg-primary h-[70px] py-6 w-[364px] bg-green-primary"
              disabled=""
              type="submit"
            >
              Update
            </button>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class=""
    >
      <div
        class="w-full text-black-default space-y-10"
      >
        <div
          class="grid w-full gap-2"
        >
          <div
            class="w-full space-y-5"
          >
            <label
              class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
              id="Guarantor’s Name-label"
            >
              Guarantor’s Name
            </label>
            <input
              aria-describedby=":r0:-form-item-description"
              aria-invalid="false"
              aria-labelledby="Guarantor’s Name-label"
              autocomplete="off"
              class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
              data-testid="input-element"
              id=":r0:-form-item"
              name="name"
              placeholder="Enter your guarantor’s name"
              value=""
            />
          </div>
        </div>
        <div
          class="grid w-full gap-2"
        >
          <div
            class="w-full space-y-5"
          >
            <label
              class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
              id="Job Title / Occupation-label"
            >
              Job Title / Occupation
            </label>
            <input
              aria-describedby=":r1:-form-item-description"
              aria-invalid="false"
              aria-labelledby="Job Title / Occupation-label"
              autocomplete="off"
              class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
              data-testid="input-element"
              id=":r1:-form-item"
              name="job"
              placeholder="Enter your job title / occupation"
              value=""
            />
          </div>
        </div>
        <div
          class="grid w-full gap-2"
        >
          <div
            class="w-full space-y-5"
          >
            <label
              class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-black-soft text-md"
              id="Phone No-label"
            >
              Phone No
            </label>
            <input
              aria-describedby=":r2:-form-item-description"
              aria-invalid="false"
              aria-labelledby="Phone No-label"
              autocomplete="off"
              class="font-productsans flex w-full py-4 text-lg text-black-smoky placeholder:text-gray-soft outline-none disabled:cursor-not-allowed disabled:opacity-50 bg-blue-light h-[68px] px-12 rounded-[15px]"
              data-testid="input-element"
              id=":r2:-form-item"
              name="phone"
              placeholder="Enter your phone number"
              value=""
            />
          </div>
        </div>
        <div
          class="w-full h-[70px]"
        >
          <button
            class="font-productsans inline-flex items-center justify-center whitespace-nowrap rounded-[14px] text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed text-white disabled:hover:bg-primary h-[70px] py-6 w-[364px] bg-green-primary"
            disabled=""
            type="submit"
          >
            Update
          </button>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
